File: android/gamecenterkore/GameCenterKore.cpp
#include <GameCenterKore.h>

#include <Kore/Android.h>

namespace GameCenterKore {

	int init() {
		return 0;
	}

	void showLeaderboard(const char* leaderboardID) {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "showLeaderboard", "(Ljava/lang/String;)V");
        jstring jid = env->NewStringUTF(leaderboardID);

        env->CallStaticVoidMethod(cls, methodId, jid);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void showAchievements() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "showAchievements", "()V");

        env->CallStaticVoidMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void reportScore(const char* leaderboardID, int score) {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

	    jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "reportScore", "(Ljava/lang/String;I)V");
        jstring jid = env->NewStringUTF(leaderboardID);

        env->CallStaticVoidMethod(cls, methodId, jid, score);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void reportAchievement(const char* achievementID, float percent) {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);
		
		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "reportAchievement", "(Ljava/lang/String;)V");
        jstring jid = env->NewStringUTF(achievementID);

        env->CallStaticVoidMethod(cls, methodId, jid);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void resetAchievements() {
		
	}



	void startQuickGame() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "startQuickGame", "()V");

        env->CallStaticVoidMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void startInviteGame() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "startInviteGame", "()V");

        env->CallStaticVoidMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void showInvitations() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "showInvitations", "()V");

        env->CallStaticVoidMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	void sendReliableDataToOthers(int data) {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "sendReliableDataToOthers", "(I)V");

        env->CallStaticVoidMethod(cls, methodId, data);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	int getReceivedData() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "getReceivedData", "()I");

        int result = env->CallStaticIntMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();

        return result;
	}

	bool getGameStarted() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "getGameStarted", "()Z");

        bool result = env->CallStaticBooleanMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();

        return result;
	}

	const char* getLocalParticipantId() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "getLocalParticipantId", "()Ljava/lang/String;");

        jstring returnString = (jstring)env->CallStaticObjectMethod(cls, methodId);
        const char *result = env->GetStringUTFChars(returnString, NULL);
        //env->ReleaseStringUTFChars(returnString, result);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();

        return result;
	}

	const char* getOpponentParticipantId() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "getOpponentParticipantId", "()Ljava/lang/String;");

        jstring returnString = (jstring)env->CallStaticObjectMethod(cls, methodId);
        const char *result = env->GetStringUTFChars(returnString, NULL);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();

        return result;
	}

	void leaveRoom() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "leaveRoom", "()V");

        env->CallStaticVoidMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();
	}

	bool getGameDisconnected() {
		JNIEnv* env;
		KoreAndroid::getActivity()->vm->AttachCurrentThread(&env, NULL);

		jclass cls = KoreAndroid::findClass(env, "gamecenterkore.GameCenterKore");

        jmethodID methodId = env->GetStaticMethodID(cls, "getGameDisconnected", "()Z");

        bool result = env->CallStaticBooleanMethod(cls, methodId);

        KoreAndroid::getActivity()->vm->DetachCurrentThread();

        return result;
	}
}

File: android/java/gamecenterkore/GameCenterKore.java
package gamecenterkore;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.WindowManager;

import com.google.android.gms.games.Games;
import com.google.android.gms.games.multiplayer.Participant;
import com.google.android.gms.games.multiplayer.realtime.RoomConfig;
import com.ktxsoftware.kore.KoreActivity;

public class GameCenterKore {
    
	public static void showLeaderboard(String leaderboardID) {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		// inst.startActivityForResult(
		// 	Games.Leaderboards.getLeaderboardIntent(inst.mGoogleApiClient, leaderboardID),
		// 	100/*REQUEST_LEADERBOARD*/);

		inst.startActivityForResult(
			Games.Leaderboards.getAllLeaderboardsIntent(inst.mGoogleApiClient),
			100/*REQUEST_LEADERBOARD*/);
	}

	public static void showAchievements() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		inst.startActivityForResult(
			Games.Achievements.getAchievementsIntent(inst.mGoogleApiClient),
        	200/*REQUEST_ACHIEVEMENTS*/);
	}

	public static void reportScore(String leaderboardID, int score) {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		Games.Leaderboards.submitScore(inst.mGoogleApiClient, leaderboardID, score);
	}

	public static void reportAchievement(String achievementID) {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		Games.Achievements.unlock(inst.mGoogleApiClient, achievementID);
	}


	public static void startQuickGame() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		// auto-match criteria to invite one random automatch opponent.
	    // You can also specify more opponents (up to 3).
	    Bundle am = RoomConfig.createAutoMatchCriteria(1, 1, 0);

	    // build the room config:
	    RoomConfig.Builder roomConfigBuilder = makeBasicRoomConfigBuilder(inst);
	    roomConfigBuilder.setAutoMatchCriteria(am);
	    roomConfigBuilder.setVariant(1);
	    RoomConfig roomConfig = roomConfigBuilder.build();

	    // create room:
	    Games.RealTimeMultiplayer.create(inst.mGoogleApiClient, roomConfig);

	    // prevent screen from sleeping during handshake
		//inst.runOnUiThread(new Runnable() {
		//	@Override
		//	public void run() {
		//	}
		//});
	    //inst.getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

	    // go to game screen
	}

	public static void startInviteGame() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;
		
		// launch the player selection screen
		// minimum: 1 other player; maximum: 1 other player
		Intent intent = Games.RealTimeMultiplayer.getSelectOpponentsIntent(inst.mGoogleApiClient, 1, 1);
		inst.startActivityForResult(intent, 10000); // RC_SELECT_PLAYERS
	}

	public static void showInvitations() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;
		
		// launch the intent to show the invitation inbox screen
		Intent intent = Games.Invitations.getInvitationInboxIntent(inst.mGoogleApiClient);
		inst.startActivityForResult(intent, 10001); // RC_INVITATION_INBOX
	}

	// create a RoomConfigBuilder that's appropriate for your implementation
	private static RoomConfig.Builder makeBasicRoomConfigBuilder(KoreActivity inst) {
		return RoomConfig.builder(inst)
				.setMessageReceivedListener(inst)
				.setRoomStatusUpdateListener(inst);
	}

	public static int getReceivedData() {
		KoreActivity inst = KoreActivity.getInstance();

		int result = inst._receivedData;
		inst._receivedData = 0;
		return result;
	}

	public static boolean getGameStarted() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return false;

		boolean result = inst._gameStarted;
		inst._gameStarted = false;
		return result;
	}

	static byte[] mMsgBuf = new byte[1];

	public static void sendReliableDataToOthers(int data) {
		mMsgBuf[0] = (byte)(data & 0xFF);

		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		for (Participant p : inst.mParticipants) {
		    if (!p.getParticipantId().equals(inst.mMyId)) {
		        Games.RealTimeMultiplayer.sendReliableMessage(inst.mGoogleApiClient, null, mMsgBuf,
		                inst.mRoomId, p.getParticipantId());
		    }
		}
	}

	public static String getLocalParticipantId() {
		KoreActivity inst = KoreActivity.getInstance();
		return inst.mMyId;
	}

	public static String getOpponentParticipantId() {
		KoreActivity inst = KoreActivity.getInstance();
		return inst.mOpponentId;
	}

	public static void leaveRoom() {
		KoreActivity inst = KoreActivity.getInstance();

		if (!inst.mGoogleApiClient.isConnected()) return;

		if (inst.mRoomId != null) {
			Games.RealTimeMultiplayer.leave(inst.mGoogleApiClient, inst, inst.mRoomId);
		}
	}

	public static boolean getGameDisconnected() {
		KoreActivity inst = KoreActivity.getInstance();

		boolean result = inst._gameDisconnected;
		inst._gameDisconnected = false;
		return result;
	}
}
